napari/components/layerlist.py

```diff
@@ -67,13 +69,58 @@ def insert(self, index: int, value: Layer):
         """Insert ``value`` before index."""
         new_layer = self._type_check(value)
         new_layer.name = self._coerce_name(new_layer.name)
+        new_layer.events.select.connect(self._on_layer_selected_change)
+        new_layer.events.deselect.connect(self._on_layer_selected_change)
         super().insert(index, new_layer)
+        self._on_layer_selected_change()
+
+    def __delitem__(self, key):
+        # Disconnect events
+        for _, index in self._delitem_indices(key):
+            disconnect_events(self[index].events, self)
+        super().__delitem__(key)
+        self._on_layer_selected_change()
 
     @property
     def selected(self):
         """List of selected layers."""
         return [layer for layer in self if layer.selected]
 
+    @property
+    @lru_cache(maxsize=12)
+    def active(self):
```

Comment:
I'm very curious @tlambert03 how you see something like this fitting in with the layer group stuff. Do layer groups have a `selected` property if all their children are selected? Do we still have an `active` concept? I'd rather not add this piece of API if we will take it away again immediately when we add layer groups, but i also want to move away from `viewer.active_layer`. What I like about this approach is that it is a dynamically calculated property that is cached instead of an attribute, but maybe that difference is small

Reply:
> Do layer groups have a selected property if all their children are selected?

no, for selection purposes, layer groups would probably be just another layer.  They take up their own "row" on the tree view, they probably wouldn't become indirectly selected by nature of selecting everything below (it would be hard, in that case, to apply changes to everything in a layer group without modifying the group itself). 

> Do we still have an active concept?

I feel like unless we have good reason not to, we should keep it.  It closely resembles the [QtSelectionModel](https://doc.qt.io/qt-5/qitemselectionmodel.html), which has both the concepts of "[selected](https://doc.qt.io/qt-5/qitemselectionmodel.html#selectedIndexes)" indices (plural), as well as a single optional "[active](https://doc.qt.io/qt-5/qitemselectionmodel.html#currentIndex)" index.

> I'd rather not add this piece of API if we will take it away again immediately when we add layer groups,

I don't think the concept will disappear, but I do think it will change.  One immediately noticeable thing is that the concept of "active layer" has moved, but the concept of "selected layers" hasn't moved.  Ultimately, I think those two things should live at the same level, on the same object.  perhaps the layerlist itself, perhaps a selection model like `layers.selection.current` and `layers.selection.active`

> What I like about this approach is that it is a dynamically calculated property that is cached instead of an attribute,

I was thinking that would also be reversed such that the selection model that doesn't have to ask each member if it's selected, but _knows_ what is selected.  (and if a layer wants to know whether it is selected, it asks the list, rather than vice versa).  conceptually, that allows a layer to have a different selection state in two different lists (perhaps something that may never happen, but "feels" right)

Reply:
Thanks for the helpful info on plans for the layer groups, that all sounds good.

> One immediately noticeable thing is that the concept of "active layer" has moved, but the concept of "selected layers" hasn't moved.

Ah yes, agreed, I thought we'd do that in its own PR.

> Ultimately, I think those two things should live at the same level, on the same object. perhaps the layerlist itself, perhaps a selection model like layers.selection.current and layers.selection.active

I was thinking just `layers.selected` and `layers.active`, but that's an interesting idea too. For `layers.selected` unfortunately it sort of needs to be an `EventedSet` which is a concept we don't really have yet, but might need to make.

> I was thinking that would also be reversed such that the selection model that doesn't have to ask each member if it's selected, but knows what is selected. (and if a layer wants to know whether it is selected, it asks the list, rather than vice versa). conceptually, that allows a layer to have a different selection state in two different lists (perhaps something that may never happen, but "feels" right)

Hmm interesting, I guess I had been thinking that a layer wouldn't then know about it's selected state, but we could revisit this.

-----

All in all though it seems like this is enough of a step in the right direction such that even if we don't move the `layer.selected` in this PR, dropping `viewer.active_layer` is still good.

I'd then say this is ready for detailed review and let's try and get it in close to where it is at now

Reply:
did `lru_cache` is a good idea? There is a hash function for this class? How cache should know that it is outdated?

Reply:
he's invalidating the cache [here](https://github.com/napari/napari/pull/1952/files#diff-acfb43cfe7047ea39e1d0f252aa2225d6a7f3f2f1fd59e11efb7851c219427d1R123) in line 123.  However, this is a perfect application for [`cached_property`](https://docs.python.org/dev/library/functools.html?#functools.cached_property), which is only in 3.8+ ... that's a useful and somewhat small function, I think we should consider vendoring it and using in 3.7

Reply:
But in such a situation why use `cache_size=12`? It should be `cache_size=1`or maybe `cache_size=2`.

Reply:
it should be 1.  there will only ever be 1 "parameter" set.  (that's why cached_property makes the most sense here)

Reply:
> Ah yes, agreed, I thought we'd do that in its own PR.

yep, that's fine.


>I unfortunately it sort of needs to be an `EventedSet` which is a concept we don't really have yet, but might need to make.

"unfortunately"??  that sounds like fun! :joy:  I have an "evented dict" actually that I never pushed because we didn't need it yet.  but I think it'd be great to fill out the `events.containers` folder.

> if a layer wants to know whether it is selected, it asks the list, rather than vice versa
> > Hmm interesting, I guess I had been thinking that a layer wouldn't then know about it's selected state, but we could revisit this.

my point wasn't so much that a layer _should_ need to know about it's selected state, but rather that that a layer shouldn't have a `layer.selected` attribute, and _if_ the layer (or anything else) needs to know whether that layer is selected, it asks the list or the selection model (not the layer)


> All in all though it seems like this is enough of a step in the right direction such that even if we don't move the `layer.selected` in this PR, dropping `viewer.active_layer` is still good.
> 
> I'd then say this is ready for detailed review and let's try and get it in close to where it is at now

I guess my main hesitation here is the addition of new `layerlist.events.active` event.  That's where we're starting to build out a new API, and where it would be nice to know where this whole extraction of the selection model is going to land.  So while I'm 100% on board with getting this stuff off the viewer, I do find myself wondering about the urgency for getting this half-way PR in before we know how it's all going to settle?

Reply:
> it should be 1. there will only ever be 1 "parameter set". (that's why cached_property makes the most sense here)

I was thinking about the case where they may be multiple viewers, I believe the cache is on the class and so if you have multiple instances you need > 1. I went with 12 as I don't think we'd have more than 12 viewers.

> However, this is a perfect application for cached_property, which is only in 3.8+ ... that's a useful and somewhat small function, I think we should consider vendoring it and using in 3.7

I saw cached_property in the docs, I'm open to this, could add a note and wait till 3.8 too. I'll look closer.

> "unfortunately"?? that sounds like fun! 😂 I have an "evented dict" actually that I never pushed because we didn't need it yet. but I think it'd be great to fill out the events.containers folder.

Lol, ok, glad you're excited to do the work! Once we have that doing the `selected` set stuff will be much easier.

I'm going to respond to you last comment in main thread